### ISSUE TITLE
Add the dimension as template parameter

### ISSUE THREAD

**Issue by hugandr** [PatWie/CppNumericalSolvers#41]
I find it often very confusing and inconvenient that the dimension is not a template argument of the problem. Wouldn't it make sense to add it there?

Then you don't always have to find out the dimension dynamically.

**Comment by PatWie** [comment]
As I wrote in the other comment. This sounds good to me and this is the way I would suggest to do.

**Comment by tesch1** [comment]
A while ago I made some of the changes that you guys are discussing - templated size params, auto bounds when bounds arent specified (#40) -- probably it's not mergeable into the current state of the ccode anymore, but it might help as a reference for how some features could be implemented.  (Also there are asa_cg and ipopt wrappers and a mechanism to dynamically instantiate different solvers on the same problem, and support for auto-differentiation) ... 
https://github.com/tesch1/CppNumericalSolvers/tree/templated/src

**Comment by spinicist** [comment]
I have made an attempt at this in PR #43 (excuse the name, BoundedProblem is next on the TODO list).

This is quite a major refactor and touches a lot of files, so comments are welcomed. In most situations it allows users to use either fixed-size or dynamic-sized problems. However for certain optimizers (LBFGSB) I think currently you will get nasty run-time errors if you try with a dynamic-sized problem.

@tesch1, I took a quick look at your repo and the work is impressive. However I agree that it has diverged a bit too far from where master is at the moment to merge it. Feel free to make comments on where I'm going with my ideas though!
